src/main/java/net/imglib2/algorithm/localextrema/LocalExtrema.java

```diff
@@ -92,79 +100,370 @@
+	 */
+	public static < P, T > ArrayList< P > findLocalExtrema(
+			final RandomAccessible< T > source,
+			final Interval interval,
+			final LocalNeighborhoodCheck< P, T > localNeighborhoodCheck,
+			final Shape shape,
+			final ExecutorService service,
+			final int numTasks )
+	{
+
+		final int nDim = source.numDimensions();
+
+		final long[] min = Intervals.minAsLongArray( interval );
+		final long[] max = Intervals.maxAsLongArray( interval );
 
-		final ArrayList< Future< Void > > futures = new ArrayList< Future< Void > >();
-		final List< P > synchronizedAllExtrema = Collections.synchronizedList( allExtrema );
-		for ( int taskNum = 0; taskNum < numTasks; ++taskNum )
+		final int maxSizeDim = IntStream.range( 0, nDim ).mapToObj( d -> new ValuePair<>( d, interval.dimension( d ) ) ).max( ( p1, p2 ) -> Long.compare( p1.getB(), p2.getB() ) ).get().getA();
+		final long maxDimSize = interval.dimension( maxSizeDim );
+		final long maxDimMax = max[ maxSizeDim ];
+		final long maxDimMin = min[ maxSizeDim ];
+		final long taskSize = Math.max( maxDimSize / numTasks, 1 );
+
+		final ArrayList< Callable< ArrayList< P > > > tasks = new ArrayList<>();
+
+		for ( long start = maxDimMin, stop = maxDimMin + taskSize - 1; start <= maxDimMax; start += taskSize, stop += taskSize )
 		{
-			min[ splitd ] = full.min( splitd ) + taskNum * dsize;
-			max[ splitd ] = ( taskNum == numTasks - 1 ) ? full.max( splitd ) : min[ splitd ] + dsize - 1;
-			final RandomAccessibleInterval< T > source = Views.interval( img, new FinalInterval( min, max ) );
-			final ArrayList< P > extrema = new ArrayList< P >( 128 );
-			final Callable< Void > r = new Callable< Void >()
-			{
-				@Override
-				public Void call()
-				{
-					final Cursor< T > center = Views.flatIterable( source ).cursor();
-					for ( final Neighborhood< T > neighborhood : shape.neighborhoods( source ) )
-					{
-						center.fwd();
-						final P p = localNeighborhoodCheck.check( center, neighborhood );
-						if ( p != null )
-							extrema.add( p );
-					}
-					synchronizedAllExtrema.addAll( extrema );
-					return null;
-				}
-			};
-			futures.add( service.submit( r ) );
+			final long s = start;
+			// need max here instead of dimension for constructor of
+			// FinalInterval
+			final long S = Math.min( stop, maxDimMax );
+			tasks.add( () -> {
+				final long[] localMin = min.clone();
+				final long[] localMax = max.clone();
+				localMin[ maxSizeDim ] = s;
+				localMax[ maxSizeDim ] = S;
+				return findLocalExtrema( source, new FinalInterval( localMin, localMax ), localNeighborhoodCheck, shape );
+			} );
 		}
-		for ( final Future< Void > f : futures )
+
+		final ArrayList< P > extrema = new ArrayList<>();
+		// TODO It is probably better to throw exception than to use try/catch
+		// block and return potentially incomplete/inconsistent list of extrema.
+		// Returning an empty list on exception could be a compromise without
+		// changing the interface.
```

Comment:
Sounds reasonable to me! I will add that. 

As we are at interface changes already, what about @tinevez suggestion in #33 :
https://github.com/imglib/imglib2-algorithm/pull/33#pullrequestreview-25614889
We could return a `List` in the new interface but cast to `ArrayList` before returning in the deprecated version because we know the implementation details.
I still think that we should stick with `ArrayList` because that is unlikely to change but I'd like to hear your opinion on it.